repos/system_upgrade/common/actors/missinggpgkeysinhibitor/libraries/missinggpgkey.py

```diff
@@ -0,0 +1,354 @@
+    """
+    if not output or output['exit_code']:
+        return []
+
+    # we are interested in the lines of the output starting with "pub:"
+    # the colons are used for separating the fields in output like this
+    # pub:-:4096:1:999F7CBF38AB71F4:1612983048:::-:::escESC::::::23::0:
+    #              ^--------------^ this is the fingerprint we need
+    #                      ^------^ but RPM version is just the last 8 chars lowercase
+    # Also multiple gpg keys can be stored in the file, so go through all "pub"
+    # lines
+    gpg_fps = []
+    for line in output['stdout']:
+        if not line or not line.startswith('pub:'):
+            continue
+        parts = line.split(':')
+        if len(parts) >= 4 and len(parts[4]) == 16:
+            gpg_fps.append(parts[4][8:].lower())
+        else:
+            api.current_logger().warning(
+                'Cannot parse the gpg2 output. Line: "{}"'
+                .format(line)
+            )
+
+    return gpg_fps
+
+
+def _read_gpg_fp_from_file(key_path):
+    """
+    Returns the list of public key fingeprints in the given file
+
+    Logs warning in case no OpenPGP data found in the given file or it is not
+    readable for some reason.
+    """
+    res = _gpg_show_keys(key_path)
+    fp = _parse_fp_from_gpg(res)
+    if not fp:
+        error = 'Unable to read OpenPGP keys from {}: {}'.format(key_path, res['stderr'])
+        api.current_logger().error(error)
+    return fp
+
+
+def _get_path_to_gpg_certs():
+    """
+    Get path to the directory with trusted target gpg keys
+    """
+    target_major_version = get_target_major_version()
+    target_product_type = config.get_product_type('target')
+    certs_dir = target_major_version
+    # only beta is special in regards to the GPG signing keys
+    if target_product_type == 'beta':
+        certs_dir = '{}beta'.format(target_major_version)
+    return os.path.join(api.get_common_folder_path(GPG_CERTS_FOLDER), certs_dir)
+
+
+def _expand_vars(path):
+    """
+    Expand variables like $releasever and $basearch to the target system version
+    """
+    r = path.replace('$releasever', get_target_major_version())
+    r = r.replace('$basearch', api.current_actor().configuration.architecture)
+    return r
+
+
+def _get_abs_file_path(target_userspace, file_url):
+    """
+    Return the absolute path for file_url if starts with file:///
+
+    If the file_url starts with 'file:///', return its absolute path to
+    the target userspace container, as such a file is supposed to be located
+    on the target system.
+
+    For all other cases, return the originally obtained value.
+    """
+    # TODO(pstodulk): @Jakuje: are we sure the file will be inside the
+    # target userspace container? What if it's a file locally stored by user
+    # and the repository is defined like that as well? Possibly it's just
+    # a corner corner case. I guess it does not have a high prio tbh, but want
+    # to be sure.
```

Comment:
Thinking about this one a bit more, it might make sense to create a fallback in case the file is not in the target userspace container. It will be there only if it is provided by some of the packages that were installed initially together with the rpm/dnf and stuff requested by customer. In case it is just random file dropped in the filesystem (alongside with the random repo files), it will be only in the source filesystem and we will have to look there. I think this will be more common case and if this would not be that complicated, I would propose to log something here and fall-back.

Reply:
yes. that would be probably best